FeatureEngine.py

```diff
@@ -42,37 +44,243 @@ def map_labels_to_invoice_OCR(self, data_path: str, verbose: bool = False):
+        features["contains_total_label"] = True if token.total_label else False
+        features["contains_date_label"] = True if token.date_label else False
+        features["contains_digit"] = True if token.contains_digit else False
 
-        return features
+        # boolean if aligned with selected tokens
+        moe = 10  # arbitary 10 pixle margin of error
+        features["vert_align_to_cell_w_date"] = False
+        features["vert_align_to_cell_w_currency"] = False
+        features["vert_align_to_cell_w_address"] = False
+        features["vert_align_to_cell_w_datelabel"] = False
+        features["vert_align_to_cell_w_numlabel"] = False
+        features["vert_align_to_cell_w_totallabel"] = False
+        features["vert_align_to_cell_w_digit"] = False
+
+        for t in invoicePage.grouped_tokens:
+            if is_vert_aligned(t, token, moe):
+                if t.date_values:
+                    features["vert_align_to_cell_w_date"] = True
+                if t.currency:
+                    features["vert_align_to_cell_w_currency"] = True
+                if t.address:
+                    features["vert_align_to_cell_w_address"] = True
+                if t.date_label:
+                    features["vert_align_to_cell_w_datelabel"] = True
+                if t.num_lable:
+                    features["vert_align_to_cell_w_numlabel"] = True
+                if t.total_label:
+                    features["vert_align_to_cell_w_totallabel"] = True
+                if t.contains_digit:
+                    features["vert_align_to_cell_w_digit"] = True
 
-    ####### Functions that help create features go here #######
+            if is_hori_aligned(t, token, moe):
+                if t.date_values:
+                    features["hori_align_to_cell_w_date"] = True
+                if t.currency:
+                    features["hori_align_to_cell_w_currency"] = True
+                if t.address:
+                    features["hori_align_to_cell_w_address"] = True
+                if t.date_label:
+                    features["hori_align_to_cell_w_datelabel"] = True
+                if t.num_lable:
+                    features["hori_align_to_cell_w_numlabel"] = True
+                if t.total_label:
+                    features["hori_align_to_cell_w_totallabel"] = True
+                if t.contains_digit:
+                    features["hori_align_to_cell_w_digit"] = True
 
+        # dist to nearest cell with field (inf if no field in page)
+        features["dist_nearest_cell_w_date"] = math.inf
+        features["dist_nearest_cell_w_currency"] = math.inf
+        features["dist_nearest_cell_w_address"] = math.inf
+        features["dist_nearest_cell_w_datelabel"] = math.inf
+        features["dist_nearest_cell_w_numlabel"] = math.inf
+        features["dist_nearest_cell_w_totallabel"] = math.inf
+        features["dist_nearest_cell_w_digit"] = math.inf
+
+        for t in invoicePage.grouped_tokens:
```

Comment:
Same as above. grouped_tokens would include the token that is currently being examined, so distance would be 0 if it contains the field

Reply:
solved